### PR TITLE
Bootstrapnas - Elastic Width Configuration

### PR DESCRIPTION
**Pull request by jpablomch**
### Changes

Improved mechanisms to specify a desired width for a super-network. 

### Reason for changes

More control when generating super-networks.

### Tests

Modified the previously included tests for elastic width to support the improvements in this PR.

### CODE REVIEW

File: nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_depth.py [hunk 1]
```diff
@@ -397,6 +397,8 @@ def build(self, target_model: NNCFNetwork) -> ElasticDepthHandler:
                 end_node = graph.get_node_by_name(block.end_node_name)
                 ordinal_ids.append([start_node.node_id, end_node.node_id])
             self._ordinal_ids = ordinal_ids
+            str_bs = [str(block) for block in self._skipped_blocks]
+            nncf_logger.info('\n'.join(['\n\"Manual depth blocks:\": [', ',\n'.join(str_bs), ']']))
```
Comment (vinnamkim) [inline nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_depth.py:401]: How about using f-string? e.g.

```
f"""
Manual depth blocks:
{str_bs}
"""
```

File: nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py [hunk 2]
```diff
@@ -204,19 +217,25 @@ def get_state(self) -> Dict[str, Any]:
...
+        return f"{self.__class__.__name__}: width_step: {self.width_step} min_width: {self.min_width} width_multipliers: {self.width_multipliers} max_num_widths: {self.max_num_widths} overwrite_groups: {self.overwrite_groups} overwrite_group_widths: {self.overwrite_groups_widths}"
+
 
 class ElasticOutputWidthOp(ElasticWidthOp):
     """
     Base class for trimming output channels (output width) of the operations.
     """
 
-    def __init__(self, max_width: int, node_name: str, params: ElasticWidthParams):
+    def __init__(self, max_width: int, node_name: str, params: ElasticWidthParams, fixed_width_list=[]):
```
Comment (vinnamkim) [inline nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py:238]: Would you please add a type hint for consistency?

File: nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py [hunk 3]
```diff
@@ -427,7 +456,8 @@ def __init__(self, target_model: NNCFNetwork,
                  weights_normalizer_fn: Optional[Callable[[torch.Tensor], torch.Tensor]],
                  node_name_vs_dynamic_input_width_op_map: Dict[NNCFNodeName, ElasticWidthOp],
                  pruned_module_groups_info: Clusterization[ElasticWidthInfo](id_fn=lambda x: x.node_name),
-                 transformation_commands: List[TransformationCommand]):
+                 transformation_commands: List[TransformationCommand],
+                 add_dynamic_inputs = None):
```
Comment (vinnamkim) [inline nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py:460]: Is this variable type `List[str]` or `Optional[List[str]]`?

File: nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py [hunk 4]
```diff
@@ -841,7 +905,13 @@ def build(self, target_model: NNCFNetwork) -> ElasticWidthHandler:
...
                     raise RuntimeError(f'Elastic width is not supported for {metatype}')
                 elastic_op_creator = metatype_vs_elastic_op_creator[metatype]
-                elastic_width_operation = elastic_op_creator(layer_attrs, node_name, self._params)
+
+                if self._overwriting_pruning_groups:
+                    elastic_width_operation = elastic_op_creator(layer_attrs, node_name,
+                                                                 self._params,
+                                                                 self._overwrite_groups_widths[i])
+                else:
+                    elastic_width_operation = elastic_op_creator(layer_attrs, node_name, self._params, [])
```
Comment (vinnamkim) [inline nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py:914]: For the better readability
```
elastic_width_operation = elastic_op_creator(
    layer_attrs, 
    node_name, 
    self._params, 
    self._overwrite_groups_widths[i] if self._overwriting_pruning_groups else []
)
```

File: nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py [hunk 5]
```diff
@@ -144,7 +147,10 @@ def __init__(self,
                  max_num_widths: int,
                  width_step: int,
                  width_multipliers: List[float],
-                 filter_importance: str):
+                 filter_importance: str,
+                 overwrite_groups: List[str],
+                 overwrite_groups_widths: List[str],
+                 add_dynamic_inputs: List[str]):
```
Comment (vinnamkim) [inline nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py:153]: Should it be `Optional[List[str]]`?

File: nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_depth.py (hunk 1 above)
Comment (jpablomch) [reply nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_depth.py]: @vinnamkim Fixed. Thanks!

File: nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py (hunk 2 above)
Comment (jpablomch) [reply nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py]: Added Type Hint. Thanks!

File: nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py (hunk 3 above)
Comment (jpablomch) [reply nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py]: @vinnamkim. Fixed. Thanks!

File: nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py (hunk 4 above)
Comment (jpablomch) [reply nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py]: @vinnamkim Fixed. Thanks!

File: nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py (hunk 5 above)
Comment (jpablomch) [reply nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py]: @vinnamkim Fixed. Thanks!

File: nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py [hunk 6]
```diff
@@ -447,9 +483,10 @@ def __init__(self, target_model: NNCFNetwork,
         self._transformation_commands = transformation_commands
         self._filter_importance_fn = filter_importance_fn
         self._weights_normalizer_fn = weights_normalizer_fn
+        self._add_dynamic_inputs = add_dynamic_inputs
 
         graph = self._target_model.get_original_graph()
-        prunable_types = [NNCFConv2d.op_func_name]
+        prunable_types = [NNCFConv2d.op_func_name] #, NNCFLinear.op_func_name]
```
Comment (vinnamkim) [inline nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py:489]: Is this you intended? I saw that you added `NNCFLinear.op_func_name` before the latest commit.

Comment (jpablomch) [reply nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py:489]: Yes, I am reverted that until I finish testing a new way to reordering the filters. Thanks!

**Review by vinnamkim — APPROVED** [review]
LGTM. But, I don't have expertise about NAS. Looks like we need more reviewers.

cc: @wonjuleee @daniil-lyakhov @negvet @sooahleex @sungchul2

File: nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_depth.py [hunk 7]
```diff
@@ -337,6 +337,9 @@ def get_state(self) -> Dict[str, Any]:
     def __eq__(self, other: 'ElasticDepthParams') -> bool:
         return self.__dict__ == other.__dict__
 
+    def __str__(self):
+        return f"{self.__class__.__name__}: max_block_size: {self.max_block_size} min_block_size: {self.min_block_size} allow_nested_blocks: {self.allow_nested_blocks} allow_linear_combinations: {self.allow_linear_combination} skipped_blocks: {self.skipped_blocks}"
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_depth.py:341]: Looks like pylint rules violation

Comment (jpablomch) [reply nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_depth.py]: @daniil-lyakhov @wonjuleee Thanks! Pylint Fixed!

**Pull request merged**